project_status_report/models/project_status_indicator.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Camptocamp (<http://www.camptocamp.com>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+
+from odoo import api, fields, models
+from odoo import _, exceptions
+from odoo import SUPERUSER_ID
+
+import sys
+import compiler
+import traceback
+
+from .common import PYTHON_CODE_DEFAULT
+
+import logging
+logger = logging.getLogger('project_status_indicator')
+
+
+class ProjectStatusIndicator(models.Model):
+    _name = 'project.status.indicator'
+
+    _order = 'sequence'
+
+    name = fields.Char(required=True, translate=True)
+    report_id = fields.Many2one(comodel_name='project.status.report')
+    sequence = fields.Integer(default=10)
+    value_type = fields.Selection(selection=[('numeric', _('Numeric')),
+                                             ('boolean', _('Boolean')),
+                                             ('text', _('Text'))
+                                             ],
+                                  required=True,
+                                  )
+    python_code = fields.Text(default=PYTHON_CODE_DEFAULT)
+
+    @api.multi
+    @api.constrains('python_code')
+    def _check_python_code_syntax(self):
+        """
+        Syntax check the python code of a step
+        """
+        for step in self:
+            try:
+                compiler.parse(step.python_code)
+            except SyntaxError, exception:
+                logger.error(''.join(traceback.format_exception(
+                    sys.exc_type,
+                    sys.exc_value,
+                    sys.exc_traceback,
+                )))
+                raise exceptions.ValidationError(
+                    _('Error in python code for step "%s"'
+                      ' at line %d, offset %d:\n%s') % (
+                        step.name,
+                        exception.lineno,
+                        exception.offset,
+                        exception.msg,
+                    ))
+
+        return True
+
+    @api.model
+    def _construct_env_dict(self, project, date, **kwargs):
+        """
+
+        """
+        analytic_account = project.analytic_account_id
+        invoices = self.env['account.invoice.line'].search(
+            [('account_analytic_id', '=', analytic_account.id)
```

Comment:
if you have tons of lines is going to be heavy.  In any case you are searching again a.a.l a couple of lines below...